nat-lab/guidelines.md

```diff
@@ -0,0 +1,215 @@
+### Natlab Guidelines:
+
+1. **Use Parametrization:**
+   - Test across all operating systems when possible.
+   - Ensure parametrization for each node.
+
+   **Why:**
+   - Covers more areas.
+   - Better readability.
+   - Wider usage of a test.
+
+   **Example:**
+   ```
+   @pytest.mark.parametrize("alpha_node", [ConnectionTag.MAC_VM, ConnectionTag.WINDOWS_VM, ConnectionTag.DOCKER_CONE_CLIENT_1])
+   @pytest.mark.parametrize("beta_node", ConnectionTag.MAC_VM)
+   def test_params(alpha_node, beta_node):
+   ```
+
+2. **Assertions Clarity:**
+   - Assertions must be explicit and easily visible in the test.
+   - Avoid assertions in different threads, keep it within test, raise an exception instead if needed.
+
+   **Why:**
+   - Easier to track which test asserted if assert is within a test.
+   - Easier to read stacktrace.
+
+3. **Start Event Checks Promptly:**
+   - Initiate event checks before the corresponding action begins.
+
+   **Why:**
+   - Starting to wait for event after action, might result in missing the event.
+
+   **Example**:
+   - DON'T:
+   ```
+      await connect_to_vpn()
+      # event might occur event before starting to wait
+      await wait_for_event_peer(State.Connected)
+   ```
+   - DO:
+   ```
+      async with asyncio_util.run_async_context(wait_for_event_peer(State.Connected)) as event:
+            await asyncio.wait_for(
+                asyncio.gather(
+                    *[
+                        connect_to_vpn()
+                        event,
+                    ]
+                ),
+                timeout
+            )
+   ```
+
+4. **Utilize Provided Helpers:**
+   - Employ provided helper functions to establish the test environment.
+
+5. **Timeouts for Potential Hangs:**
+   - Apply timeouts to actions susceptible to hanging.
+
+   **Example:**
+   - As previously mentioned, event might occur before starting to wait for it. To avoid natlab hanging, wrap coroutine or future to complete with timeout.
+   ```
+      await connect_to_vpn()
+      await asyncio.wait_for(wait_for_event_peer(State.Connected), timeout=5)
+   ```
+
+6. **VM/Docker Reversion:**
+   - Ensure any modifications to VM/Docker are reverted during test teardown.
+   - Tests should leave no residual configurations even in case of failure.
+
+   **Why:**
+   - Some tests tinker with iptables / routing tables and some other significant stuff, which might lead to failure of other tests if not properly cleaned.
```

Comment:
I think that instead of making this a guideline, such modifications can be automatically reverted using python context managers (`with` statement).

Reply:
It is already done with context managers, but I just wanted to mention that in guidelines, so that developers wouldn't forget that. Also it is not all about iptables/routing tables, I took them as an example. Other places might be changed and not restored properly

Reply:
Yeah, but the correct example you provided is not using context manager (so I assume it can permanently change the state if the exception is raised before call to `restore_tcp_conn_to_host`). I think it would be better to show in the example the best possible version :)

Reply:
Updated